test/x509_time_test.c

```diff
@@ -424,11 +424,121 @@ static int test_days(int n)
     return r;
 }
 
+static const struct {
+    ASN1_TIME asn1;
+    const char *readable;
+} x509_print_tests [] = {
+    {
+        /* Generalized Time */
+        {
+            sizeof("20170731222050Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050Z",
+            0
+        },
+        "Jul 31 22:20:50 2017 GMT"
+    },
+    {
+        /* Generalized Time, no seconds */
+        {
+            sizeof("201707312220Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"201707312220Z",
+            0
+        },
+        "Jul 31 22:20:00 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (3 digits) */
+        {
+            sizeof("20170731222050.123Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.123Z",
+            0
+        },
+        "Jul 31 22:20:50.123 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (1 digit) */
+        {
+            sizeof("20170731222050.1Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.1Z",
+            0
+        },
+        "Jul 31 22:20:50.1 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (0 digit) */
+        {
+            sizeof("20170731222050.Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.Z",
+            0
+        },
+        "Bad time value"
+    },
+    {
+        /* UTC Time */
+        {
+            sizeof("170731222050Z") - 1,
+            V_ASN1_UTCTIME,
+            (unsigned char *)"170731222050Z",
+            0
+        },
+        "Jul 31 22:20:50 2017 GMT"
+    },
+    {
+        /* UTC Time, no seconds */
+        {
+            sizeof("1707312220Z") - 1,
+            V_ASN1_UTCTIME,
+            (unsigned char *)"1707312220Z",
+            0
+        },
+        "Jul 31 22:20:00 2017 GMT"
+    }
+};
+
+static int test_x509_time_print(int idx)
+{
+    BIO *m;
+    int ret = 0, rv;
+    char *pp;
+    const char *readable;
+
+    m = BIO_new(BIO_s_mem());
+    if (m == NULL)
+        goto err;
+
+    rv = ASN1_TIME_print(m, &x509_print_tests[idx].asn1);
+    if (!TEST_int_ge(rv, 0)) {
```

Comment:
I'd roll this up: `if (!TEST_int_ge(rv = ASN1_TIME_print(m, &x509_print_tests[idx].asn1)))`
It produces a more informative error message on failure.

It's stylistic so not essential.